File: core/src/capi/signal_guard.zig
//! Signal Guard - Graceful handling of fatal signals (SIGBUS, SIGSEGV)
//!
//! Provides a mechanism to catch fatal signals and convert them to error codes
//! instead of crashing. This is used to handle resource exhaustion scenarios
//! where the kernel sends SIGBUS (e.g., mmap page fault failures).
//!
//! Usage:
//!   const guard = SignalGuard.init();
//!   const result = guard.call(myFunction, context);
//!   // result is null if SIGBUS was caught
//!
//! Thread Safety:
//!   Each thread has its own jump buffer via threadlocal storage.
//!   Multiple threads can use SignalGuard simultaneously.
//!
//! Limitations:
//!   - Only works on POSIX systems (Linux, macOS)
//!   - Cannot recover from signals raised outside call() scope
//!   - Uses sigsetjmp/siglongjmp for async-signal-safe recovery

const std = @import("std");
const builtin = @import("builtin");

// C function from signal_guard.c
// Uses callback pattern because sigsetjmp must be called at the same stack
// frame where siglongjmp will return.
extern fn talu_signal_guard_install() callconv(.c) void;
extern fn talu_signal_guard_call(
    func: *const fn (?*anyopaque) callconv(.c) c_int,
    ctx: ?*anyopaque,
) callconv(.c) c_int;

/// Signal guard for catching fatal signals.
pub const SignalGuard = struct {
    /// Initialize signal handlers. Call once per thread that needs protection.
    pub fn init() SignalGuard {
        if (comptime builtin.os.tag == .linux or builtin.os.tag == .macos) {
            talu_signal_guard_install();
        }
        return .{};
    }

    /// Cleanup (currently no-op, handlers persist for thread lifetime).
    pub fn deinit(_: SignalGuard) void {
        // Handlers remain installed - they're safe to leave in place
    }

    /// Call a function with SIGBUS protection.
    /// The callback must have signature: fn(*anyopaque) callconv(.c) c_int
    /// Returns null if SIGBUS was caught, otherwise returns the function's return value.
    pub fn call(
        _: SignalGuard,
        func: *const fn (*anyopaque) callconv(.c) c_int,
        ctx: *anyopaque,
    ) ?c_int {
        if (comptime builtin.os.tag != .linux and builtin.os.tag != .macos) {
            // No protection on unsupported platforms - just call directly
            return func(ctx);
        }

        // The C function expects fn(?*anyopaque) callconv(.c) c_int
        // We need to cast our function pointer to match
        const c_func: *const fn (?*anyopaque) callconv(.c) c_int = @ptrCast(func);
        const result = talu_signal_guard_call(c_func, ctx);
        if (result == -1) {
            return null; // Signal was caught
        }
        return result;
    }
};

// =============================================================================
// Tests
// =============================================================================

fn testCallback(ctx: *anyopaque) callconv(.c) c_int {
    const flag: *bool = @ptrCast(@alignCast(ctx));
    flag.* = true;
    return 0;
}

test "SignalGuard: basic usage without signal" {
    const guard = SignalGuard.init();
    defer guard.deinit();

    var executed = false;
    const result = guard.call(&testCallback, @ptrCast(&executed));

    try std.testing.expect(executed);
    try std.testing.expectEqual(@as(?c_int, 0), result);
}

fn testReturns42(_: *anyopaque) callconv(.c) c_int {
    return 42;
}

test "SignalGuard: return value propagation" {
    const guard = SignalGuard.init();
    defer guard.deinit();

    var dummy: bool = false;
    const result = guard.call(&testReturns42, @ptrCast(&dummy));

    try std.testing.expectEqual(@as(?c_int, 42), result);
}

// Note: SIGBUS recovery tests cannot run in the Zig test framework because
// the framework has its own signal handlers that interfere. Run via CI:
//   ./scripts/check_signal_tests.sh
//
// Or see: tests/helpers/signal/signal_guard_test.zig

File: core/src/capi/signal_guard.c
/**
 * Signal Guard - C implementation for sigsetjmp/siglongjmp
 *
 * This provides async-signal-safe recovery from SIGBUS using sigsetjmp/siglongjmp.
 * sigsetjmp/siglongjmp are macros in C, so we need this C wrapper.
 *
 * The key insight is that sigsetjmp/siglongjmp work like this:
 * - sigsetjmp saves the execution context (registers, stack pointer, etc.)
 * - siglongjmp restores that context and makes sigsetjmp return again
 *
 * The problem with wrapping sigsetjmp in a function is that when siglongjmp
 * jumps back, the function returns normally - but the caller has already moved
 * past the call site!
 *
 * Solution: Use a callback pattern where the protected code is passed as a
 * function pointer. This way, sigsetjmp and the check happen in the same
 * function, and siglongjmp correctly skips the callback.
 */

#define _POSIX_C_SOURCE 200809L
#include <signal.h>
#include <setjmp.h>
#include <stdatomic.h>
#include <stddef.h>

/* Thread-local jump buffer and state */
static _Thread_local sigjmp_buf talu_signal_guard_jmpbuf;
static _Thread_local volatile sig_atomic_t talu_signal_guard_active = 0;

/* Signal handler - jumps back to setjmp point if guard is active */
static void talu_signal_handler(int sig) {
    if (talu_signal_guard_active) {
        talu_signal_guard_active = 0;
        siglongjmp(talu_signal_guard_jmpbuf, 1);
    }
    /* If guard not active, re-raise with default handler */
    signal(sig, SIG_DFL);
    raise(sig);
}

/* Install signal handlers - call once per thread */
void talu_signal_guard_install(void) {
    struct sigaction sa;
    sa.sa_handler = talu_signal_handler;
    sigemptyset(&sa.sa_mask);
    sa.sa_flags = 0;
    sigaction(SIGBUS, &sa, NULL);
    sigaction(SIGSEGV, &sa, NULL);  /* Also catch SIGSEGV for memory access errors */
}

/*
 * Execute a function with SIGBUS protection.
 *
 * @param func The function to execute (takes void* context, returns int error code)
 * @param ctx  Context pointer passed to func
 * @return 0 if func completed successfully (func returned 0)
 *         func's return value if func returned non-zero
 *         -1 if SIGBUS was caught during execution
 */
int talu_signal_guard_call(int (*func)(void*), void* ctx) {
    if (sigsetjmp(talu_signal_guard_jmpbuf, 1) == 0) {
        talu_signal_guard_active = 1;
        int result = func(ctx);
        talu_signal_guard_active = 0;
        return result;
    } else {
        /* siglongjmp jumped here - signal was caught */
        return -1;
    }
}
